internal/util/grpcclient/client.go

```diff
@@ -311,55 +314,86 @@ func (c *ClientBase[T]) connect(ctx context.Context) error {
+		callErr   error
+		client    T
+	)
+
+	client, clientErr = c.GetGrpcClient(ctx)
+	if clientErr != nil {
+		log.Warn("fail to get grpc client", zap.Error(clientErr))
 	}
 
-	if IsCrossClusterRoutingErr(err) {
-		log.Warn("CrossClusterRoutingErr, start to reset connection", zap.Error(err))
+	resetClientFunc := func() {
 		c.resetConnection(client)
-		return ret, merr.ErrServiceUnavailable // For concealing ErrCrossClusterRouting from the client
-	}
-	if IsServerIDMismatchErr(err) {
-		log.Warn("Server ID mismatch, start to reset connection", zap.Error(err))
-		c.resetConnection(client)
-		return ret, err
+		client, clientErr = c.GetGrpcClient(ctx)
+		if clientErr != nil {
+			log.Warn("fail to get grpc client in the retry state", zap.Error(clientErr))
+		}
 	}
-	if !funcutil.CheckCtxValid(ctx) {
-		// check if server ID matches coord session, if not, reset connection
-		if c.sess != nil {
-			sessions, _, getSessionErr := c.sess.GetSessions(c.GetRole())
-			if getSessionErr != nil {
-				// Only log but not handle this error as it is an auxiliary logic
-				log.Warn("Fail to GetSessions", zap.Error(getSessionErr))
-			}
-			if coordSess, exist := sessions[c.GetRole()]; exist {
-				if c.GetNodeID() != coordSess.ServerID {
-					log.Warn("Server ID mismatch, may connected to a old server, start to reset connection", zap.Error(err))
-					c.resetConnection(client)
-					return ret, err
-				}
+
+	innerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	_ = retry.Do(innerCtx, func() error {
+		if generic.IsZero(client) {
+			log.Warn("grpc client is nil, maybe fail to get client in the retry state")
+			resetClientFunc()
+			return errors.New("empty grpc client")
+		}
+		ret, callErr = caller(client)
+		if callErr != nil {
+			if !funcutil.IsGrpcErr(callErr) && funcutil.CheckCtxValid(ctx) {
+				log.Warn("fail to grpc call because of unknown error", zap.Error(callErr))
+				// not rpc error, it will stop to retry
+				return retry.Unrecoverable(callErr)
 			}
+			// when grpc error or context error occurs, it will reset connection
+			resetClientFunc()
+			return callErr
 		}
-		// start bg check in case of https://github.com/milvus-io/milvus/issues/22435
-		go c.bgHealthCheck(client)
-		return generic.Zero[T](), err
-	}
-	if !funcutil.IsGrpcErr(err) {
-		log.Warn("ClientBase:isNotGrpcErr", zap.Error(err))
-		return generic.Zero[T](), err
+
+		var errorCode commonpb.ErrorCode
+		var reason string
+		switch res := ret.(type) {
+		case *commonpb.Status:
+			errorCode = res.GetErrorCode()
+			reason = res.GetReason()
+		case interface{ GetStatus() *commonpb.Status }:
+			errorCode = res.GetStatus().GetErrorCode()
+			reason = res.GetStatus().GetReason()
+		default:
+			// it will directly return the result
+			log.Warn("unknown return type", zap.Any("return", ret))
+			return nil
+		}
+
+		if errorCode == commonpb.ErrorCode_Success {
+			return nil
+		}
+
+		if IsCrossClusterRoutingErr(reason) || IsServerIDMismatchErr(reason) {
```

Comment:
CrossClusterRoutingErr and ServerIDMismatchErr come from error, not status.Reason